### ISSUE TITLE
Add to F-Droid

### ISSUE THREAD

**Issue by takiainen** [niedev/RTranslator#9]
F-Droid is an installable catalogue of FOSS (Free and Open Source Software) applications for the Android platform.  Any chance of adding this project to F-Droid?

https://f-droid.org/

**Comment by niedev** [comment]
In these days I will find out more about the policies of  F-Droid and update you here, because I don't know if I still can, given that NLLB is open source but not commercial, and Ml-Kit (used for language recognition in WalkieTalkie mode) is closed source, but if I can I will definitely do it.

**Comment by jermanuts** [comment]
Hey @IzzySoft, what do you think?

**Comment by IzzySoft** [comment]
MLKit is a show-stopper for F-Droid. The APK size is a show-stopper for IzzyOnDroid unfortunately, so we don't even need to raise the MLKit question there, sorry.

**Comment by niedev** [comment]
@IzzySoft Ok, no problem, maybe in the future I will reduce the APK size and find an open-source alternative to ML-Kit.

**Comment by IzzySoft** [comment]
Please let me know if you succeed with either! Especially the latter would be helping several projects. And there it would be helpful what areas the replacement(s) you found cover, so they can be matched to a project's need. Like with [Supabase](https://supabase.com/) and [appwrite](https://github.com/appwrite/appwrite) as possible replacements for Firebase. Or UnifiedPush for FCM :wink:

That said: Best luck!

**Comment by niedev** [comment]
Thank you! I will do my best

**Comment by chameleon-lizard** [comment]
I didn't look at your code yet, but aren't there any open source models for language recognition, which would make this translator absolutely free? Quick search through huggingface gave me [this](https://huggingface.co/facebook/fasttext-language-identification), wouldn't it work?

I would be happy to contribute if this module is written in python, since I don't know much java.

**Comment by niedev** [comment]
Thank you! I did some research about it a while ago, it seems I missed this one 😅, it seems very good in quality, even if a bit big (about 200M parameters), anyway, I will test it well (quality, RAM consumption, execution time) and I will evaluate whether to insert it instead of ML-Kit. The license is the same as NLLB, @IzzySoft F-Droid accept AI models with cc-by-nc licence?

As for the app, it uses only Java code and a little bit of C++ (the percentage of Python shown by GitHub statistics are from the sentencepiece library, like most of the C++ code), but the inference logic with Java is quite simple, the complex part is mainly the conversion and optimization of the neural network models in onnx format.

**Comment by IzzySoft** [comment]
> cc-by-nc licence?

The NC is the culprit there, as it violates one of the [4 essential freedoms of free software](https://en.wikipedia.org/wiki/Free_and_open-source_software#Four_essential_freedoms_of_Free_Software). So rather not I'd say.

**Comment by niedev** [comment]
Ok 👍 Thank you!

**Comment by noctux** [comment]
Hmm, I'm not sure how strict that posture is. [whoBird](https://f-droid.org/packages/org.woheller69.whobird/) is practically in the same boat with RTranslator as it seems. From the [project's github page](https://github.com/woheller69/whoBIRD):

> This app is built on the BirdNET framework by @kahst, published under CC BY NC SA 4.0 license
> At first start it downloads the BirdNet TFLite library from whoBird-TFlite, which is published under CC BY NC SA 4.0 license

Nevertheless it is available in FDroid with appropriate antifeature markers (Non-Free Assets, Non-Free Network Services). The model is not distributed with the apk but downloaded on first run.

(Please note that this is just an observation from a bystander, I've not researched whether there were any additional considerations involved during whoBIRD's admission process).

**Comment by luisvalenzuelar** [comment]
> F-Droid is an installable catalogue of FOSS (Free and Open Source Software) applications for the Android platform. Any chance of adding this project to F-Droid?
> 
> https://f-droid.org/

F-Droid policies do not allows for non-free libraries to be used. https://f-droid.org/docs/Inclusion_Policy/

**Comment by linsui** [comment]
The code has to be FOSS but the model data can be non-commercial. If MLkit is optional you can provide a flavor without it. IIUC, MLkit is only used to detect the language. So maybe you can provide a flavor which can only set the languange manually.

**Comment by niedev** [comment]
> The code has to be FOSS but the model data can be non-commercial. If MLkit is optional you can provide a flavor without it. IIUC, MLkit is only used to detect the language. So maybe you can provide a flavor which can only set the languange manually.

Thank you for the suggestion. 

Managing multiple versions of the app would complicate maintenance and I would have less time to make truly useful changes, putting the app in F-Droid is important to me, but improving the app for now is much more important.

But surely when it will be possible I will try to slowly replace all the closed source or non-commercial components with 100% open source alternatives (if they will have an equally good quality and speed).

**Comment by Ganipotes** [comment]
[Instead of f-droid](https://privsec.dev/posts/android/f-droid-security-issues/), here are some alternatives: 

[Obtainium](https://obtainium.imranr.dev): it allows you to retrieve from Github releases, so there's nothing in particular to do. It might be useful to inform users in the read-me to make updates easier.

[Accrescent](https://accrescent.app): the store is still in beta, so it's a whitelist for application submission, as it's not yet developer-friendly, but it's quite ready for users. Applications can be open source or closed source, so there's no problem with components.

IMO it would be good to add obtanium now, and keep an eye on accrescent to add it in the future when it's 1.0.

**Comment by licaon-kter** [comment]
@Ganipotes Please open a separate issue about your favorite separate solutions, other that those in the title of the issue

also, I'm under the impression that the first link magically changes _"Last updated on"_ every time I visit it, but the content is still 3 years old talking about things that were fixed long ago.

**Comment by niedev** [comment]
Hi @Ganipotes, thank you for your suggestions!
I tried Obtanium and although it is very useful it doesn't seem to be that easy to use for an average user (I myself was intimidated when I saw all those options to add an app), for now however I can add in the readme an indication on how to receive notifications for new releases on GitHub, and I was thinking in the future to integrate an update control directly into the app.
I will keep an eye on Accrescent, if it will be supported for a long time and will reach a good number of users I might add my app there too (but for now F-Droid and Play Store have priority being more mature).

**Comment by lrq3000** [comment]
I was also going to suggest Meta's fasttext model, seems like a match made in heaven with NLLB, fasttext was made from the same corpus and hence supports as many languages (hence more than Whisper).

Note that the [latest](https://github.com/facebookresearch/fairseq/blob/nllb/README.md#lid-model) language identification (LID) model supporting 217 languages is also under CC-BY-NC 4.0, hence it's "open-source for non-commercial use" (ie, non free software), but this should work for F-Droid. Direct link is here (expanded from the tinyurl): https://dl.fbaipublicfiles.com/nllb/lid/lid218e.bin

A less accurate alternative is to simply use Whisper, see the end of this collab script: https://colab.research.google.com/drive/1rS1L4YSJqKUH_3YxIQHBI982zso23wor?usp=sharing#scrollTo=Mh_e6rV62QUM

**Comment by niedev** [comment]
Hi again @lrq3000, thank you for your suggestions.

> Note that the [latest](https://github.com/facebookresearch/fairseq/blob/nllb/README.md#lid-model) language identification (LID) model supporting 217 languages is also under CC-BY-NC 4.0, hence it's "open-source for non-commercial use" (ie, non free software), but this should work for F-Droid. Direct link is here (expanded from the tinyurl): https://dl.fbaipublicfiles.com/nllb/lid/lid218e.bin

I tested this quickly and it doesn't seems to have the same quality as Google's Language identification (it failed to identify "ciao" as Italian, only increasing the length of the text works, while Google's almost never gets it wrong).

> A less accurate alternative is to simply use Whisper, see the end of this collab script: https://colab.research.google.com/drive/1rS1L4YSJqKUH_3YxIQHBI982zso23wor?usp=sharing#scrollTo=Mh_e6rV62QUM

I already tried Whisper automatic language detection but it significantly decreases the quality of speech recognition (even if the recognized language is the correct one).

**Comment by woheller69** [comment]
If automatic language detection in Walkie-Talkie mode is the only issue, I would suggest to remove it for an F-Droid version and only show the 2 language buttons at the left and right.
I would anyway prefer not to automatically start with automatic mode when entering Walkie-Talkie mode. If there is background noise or other people around are speaking it is better to press and hold the language button while speaking. And then automatic detection is not needed anyway. Otherwise all undesired stuff is translated as well which is not great on a slow phone.

**Comment by turbocool3r** [comment]
@niedev after looking at the [whoBird](https://f-droid.org/packages/org.woheller69.whobird/) app I see there is a notice in the anti-features section about the non-commercial license, but the app is still allowed into the store. I think RTranslator should be ok as well

**Comment by linsui** [comment]
No. Only non-free assets is allowed. Non-free code is not allowed.

**Comment by Francewhoa** [comment]
Suggested Open Source (Libre Source) alternatives to ML-Kit in ticket https://github.com/niedev/RTranslator/issues/166

**Comment by woheller69** [comment]
I built a version without ML Kit:
-no conversation mode (via bluetooth)
-walkie-talkie mode only manual

Otherwise seems to work well:
https://github.com/woheller69/RTranslator

Maybe F-Droid could simply patch the original source code 
(I know there are some talented ```sed -i -e ':-)))' build.gradle``` hackers over there @linsui )

**Comment by linsui** [comment]
Yep, we can also do that. :) @niedev Would that work for you? We also need fastlane structure.

**Comment by linsui** [comment]
@woheller69 Maybe it's better to add a flavor instead?

**Comment by lrq3000** [comment]
In addition to the more practical and immediate solutions offered by @linsui and @woheller69 (thank you so mucn guys !), I would like to suggest an alternative for a later horizon.

> I already tried Whisper automatic language detection but it significantly decreases the quality of speech recognition (even if the recognized language is the correct one).

Yes this is a common issue, I tried and continue to regularly try all Android implementations of Whisper, and I agreed with your observation. But this state of matter changed recently.

There is an Android app called Notely Voice that I recently found and which systematically succeeds in speech recognition using a 460MB local multiligual Whisper model. It it by far the most effective and reliable offline whisper implementation I have ever seen, and it is opensource : 
https://github.com/tosinonikute/NotelyVoice

Tip : if you articulate just a bit, the false detection rate drastically lowers, allowing for highly accurate detection when speaking in very noisy environments such as an amusement park (yes I tried for hours just in front of an attraction where people were screaming and with the music being unreasonably loud - and my phone has no noise filter, it's old). 

I tested it in both English and French and it is very reliable. Of course comparing with cloud models is always going to be unfavorable, but that it not the main "selling" point of RTranslator, it's the offline processing, otherwise for max accuracy, Google Translate or Microsoft Translator or voice LLMs such as ChatGPT and Claude are far better from end-to-end.

Now the issue is that of course this is going to slow down the conversation process for now on most current phones, but I expect this to change very fast in the next 1-2 years with the large scale deployment of phones with much bigger RAM and more performant GPU targeting improved on-device AI processing.

**Comment by woheller69** [comment]
> [@woheller69](https://github.com/woheller69) Maybe it's better to add a flavor instead?

@niedev said above
"Managing multiple versions of the app would complicate maintenance and I would have less time to make truly useful changes, putting the app in F-Droid is important to me, but improving the app for now is much more important."

Having a flavor means all that needs to be kept in sync. So maybe it is better to have an ```sed -``` hacked version now (the code has been quite stable for a while) and hopefully later - when ML Kit - has been replaced to remove the ```sed -``` stuff.

**Comment by niedev** [comment]
Hi @woheller69, @linsui, I'm already working on a new version of RTranslator (3.0) that will replace MlKit with an open source alternative, on top of many other changes, this version will be released on F-Droid, and maybe also in the Play Store. But if you guys want to make a forked version that you can release on F-Droid it is totally ok to me (changing a bit the name to clarify the distinction to the normal version, something like "RTranslator Open" or similar maybe), note that MlKit is used only in WalkieTalkie mode, so you can also keep Conversation mode.

**Comment by niedev** [comment]
@lrq3000 I tried Notely Voice, and it seems to have the same problem as my implementation when using auto detection of language, here is the transcription of the same Italian audio, with and without the auto detection:

![Image](https://github.com/user-attachments/assets/7870a21a-26ca-4850-a618-3be37589840d)

![Image](https://github.com/user-attachments/assets/b7903f9f-f448-4c25-b707-d4c409db5ef2)

However, I will analyze the code to see if they did optimizations to the performance or quality that I missed, thank you.

**Comment by woheller69** [comment]
After my changes conversation mode didn't seem to work and I didn't want to dig deeper. For most people walkie-talkie mode will probably be enough, so I simply removed the button.

**Comment by woheller69** [comment]
@linsui @niedev 
What about this idea:
As a first step F-Droid builds the app from my fork, we change app_name to "RTranslator-lite" but keep package id and icon as they are.
Once your RTranslator (3.0) is available and fully FOSS, F-Droid switches to your repo and builds the app from there. The "-lite" will be gone.

**Comment by niedev** [comment]
@woheller69 sure, it is ok to me, thank you. If in these days I will have enough time, and if it is relatively easy, I can also do a pull request to your fork to make Conversation mode work, more functionalities is always better.

**Comment by woheller69** [comment]
Build on F-Droid works:
https://gitlab.com/woheller69/fdroiddata/-/blob/RTranslator/metadata/nie.translator.rtranslator.yml
https://gitlab.com/woheller69/fdroiddata/-/jobs/12255008560

Need to modify app_name in strings and AutoName in yml
And we need a minimal fastlane structure. Then we should be able to add it to F-Droid.

**Comment by linsui** [comment]
I wound rather just wait so that we can have reproducible build.

**Comment by woheller69** [comment]
@niedev could sign the apk, then you can have reproducible build now

**Comment by linsui** [comment]
If @niedev would like to do that I have no problem. :)

**Comment by niedev** [comment]
Sure, in these days I will do it (I don't have a PC with me right now)

**Comment by woheller69** [comment]
Fastlane is done
Name changed to ```RTranslator Open```
 
F-Droid build also works.
So now @niedev can try if the conversation mode can be fixed in a simple way and then we need to get into reproducible build, where I have no experience.

**Comment by linsui** [comment]
@woheller69 Just add Binaries and AllowedAPKSigningKeys and we'll see if there is and reproducible build issue.

**Comment by woheller69** [comment]
AllowedAPKSigningKeys is the SHA256 fingerprint of the key?

**Comment by linsui** [comment]
Yes, you can see it with apksigner verify --print-certs.

**Comment by woheller69** [comment]
@linsui I think we need some help.

I added the link to the binary and the SHA-256 of the certificate.

https://gitlab.com/woheller69/fdroiddata/-/blob/RTranslator/metadata/nie.translator.rtranslator.yml

Build works, but verification fails:

```
2025-12-04 13:13:41,144 DEBUG: tmp/binaries/nie.translator.rtranslator_24.binary.apk: Verifies
Verified using v1 scheme (JAR signing): false
Verified using v2 scheme (APK Signature Scheme v2): true
Verified using v3 scheme (APK Signature Scheme v3): false
Verified using v4 scheme (APK Signature Scheme v4): false
Verified for SourceStamp: false
Number of signers: 1
2025-12-04 13:13:41,327 INFO: ...NOT verified - /tmp/tmp502pud04/sigcp_nie.translator.rtranslator_24.apk
```

https://gitlab.com/woheller69/fdroiddata/-/jobs/12322819006

**Comment by woheller69** [comment]
modified the yml, so it now builds from @niedev own fork https://github.com/LucaMartino97/RTranslator, but the 
verification issue remains...

**Comment by linsui** [comment]
Apply the fix in https://f-droid.org/docs/Reproducible_Builds/#ndk-hash-style and https://f-droid.org/docs/Reproducible_Builds/#ndk-build-id to the .so files. Apply the fix in https://f-droid.org/docs/Reproducible_Builds/#bug-line-ending-differences-between-windows-and-linux-builds to fix the xml files.

**Comment by woheller69** [comment]
@niedev do you want reproducible build, so F-Droid publishes the app with your signature?
Then this needes to be fixed. Otherwise F-Droid can just build and sign with their own key.

**Comment by niedev** [comment]
@woheller69 I've read that I can add reproducible build in the future, so for now the second option is ok for me.

**Comment by woheller69** [comment]
The issue is that people who installed the app already will not be able to upgrade to newer versions anymore because the signature is different if this is changed in future...

**Comment by niedev** [comment]
Ok, I'm working on the 3.0 version and I don't have time now to fix these issues myself. At this point I would say that we have 2 options:

1) You make a true fork of the app, changing the package name so you can manage the app however you want and publish it as RTranslator Open yourself. This way we can also solve update issues because they will be separate apps.

2) (easier for you) We wait few months for the release of RTranslator 3.0 and I will manage the release on F-Droid myself.

Let me know if you have other options in mind.

**Comment by linsui** [comment]
It's not an issue that will take you much time to fix.  If @woheller69 can fix that @niedev you can merge the fix.

@woheller69 Do you want to fix the reproducible build issue?

**Comment by woheller69** [comment]
for good reason I don`t waste my time on reproducible build for my own apps. I don't think I want to dig deeper into this topic.
Maybe someone else wants to do it...

**Comment by linsui** [comment]
Then will you just fork it and submit to F-Droid?

**Comment by woheller69** [comment]
Then we need to change app id, icon, name, fastlane,...
So better wait

**Comment by KaKi87** [comment]
Hi,
How's this going ?
Thanks
